code/game/atoms.dm

```diff
@@ -390,24 +390,54 @@
 
 /atom/proc/build_reagent_description(mob/user)
 	. = list()
-	if(reagents)
+	if(!reagents)
+		return
+	var/one_percent = reagents.total_volume / 100
+	var/blood_type = ""
+	if(user.advanced_reagent_vision())	// You can see absolute unit concentrations in transparent containers and % concentrations in opaque containers. You can also see blood types.
+		. += "<span class='notice'>It contains:</span>"
+		if(!length(reagents.reagent_list))
+			. += "<span class='notice'>Nothing.</span>"
+			return
 		if(container_type & TRANSPARENT)
-			. += "<span class='notice'>It contains:</span>"
-			if(length(reagents.reagent_list))
-				if(user.can_see_reagents()) //Show each individual reagent
-					for(var/I in reagents.reagent_list)
-						var/datum/reagent/R = I
-						. += "<span class='notice'>[R.volume] units of [R.name]</span>"
-				else //Otherwise, just show the total volume
-					if(reagents && length(reagents.reagent_list))
-						. += "<span class='notice'>[reagents.total_volume] units of various reagents.</span>"
-			else
-				. += "<span class='notice'>Nothing.</span>"
-		else if(container_type & AMOUNT_VISIBLE)
-			if(reagents.total_volume)
-				. += "<span class='notice'>It has [reagents.total_volume] unit\s left.</span>"
+			for(var/I in reagents.reagent_list)
+				var/datum/reagent/R = I
+				if(R.id != "blood")
+					. += "<span class='notice'>[R.volume] units of [R] ([round(R.volume / one_percent)]%)</span>"
+				else
+					blood_type = R.data["blood_type"]
+					. += "<span class='notice'>[R.volume] units of [R] ([blood_type ? "[blood_type]" : ""]) ([round(R.volume / one_percent)]%)</span>"
+			return
+		// Opaque containers
+		for(var/datum/reagent/R in reagents.reagent_list)
+			if(R.id != "blood")
+				. += "<span class='notice'>[R] ([round(R.volume / one_percent)]%)</span>"
 			else
-				. += "<span class='danger'>It's empty.</span>"
+				blood_type = R.data["blood_type"]
+				. += "<span class='notice'>[R] ([blood_type ? "[blood_type]" : ""]) ([round(R.volume / one_percent)]%)</span>"
+		return
+
+	if(container_type & TRANSPARENT)
+		. += "<span class='notice'>It contains:</span>"
+		if(user.reagent_vision())	// You can see absolute unit quantities of reagents in transparent containers.
+			if(container_type & TRANSPARENT)
```

Comment:
You need to check if its transparent twice here?

Reply:
Oops! Fixed.